File: sql/plsql/JavaScript/PlSqlParserBase.js
import antlr4 from 'antlr4';
import JavaScriptLexer from './PlSqlParser.js';
import PlSqlLexer from './PlSqlLexer.js';

export default class PlSqlParserBase extends antlr4.Parser {
  _isVersion10;
  _isVersion11;
  _isVersion12;
  _lastUnitWasPlsql;

  constructor(input) {
    super(input);
    this._isVersion10 = true;
    this._isVersion11 = true;
    this._isVersion12 = true;
    this._lastUnitWasPlsql = false;
  }

  reset() {
    this._lastUnitWasPlsql = false;
    super.reset();
  }

  setLastUnitPlsql() { this._lastUnitWasPlsql = true; }
  setLastUnitSql()   { this._lastUnitWasPlsql = false; }
  isLastUnitSql()    { return !this._lastUnitWasPlsql; }
  isLastUnitPlsql()  { return this._lastUnitWasPlsql; }

  isSolidusSeparator() {
    const stream = this.getTokenStream();
    const solidus = stream.LT(1);
    if (solidus == null || solidus.type !== PlSqlLexer.SOLIDUS)
      return false;

    const solidusLine = solidus.line;

    const prev = stream.LT(-1);
    if (prev != null && prev.type !== antlr4.Token.EOF && prev.line === solidusLine)
      return false;

    const next = stream.LT(2);
    if (next != null && next.type !== antlr4.Token.EOF && next.line === solidusLine)
      return false;

    return true;
  }

  isVersion10() {
    return this._isVersion10;
  }

  isVersion11() {
    return this._isVersion11;
  }

  isVersion12() {
    return this._isVersion12;
  }

  setVersion10(value) {
    this._isVersion10 = value;
  }

  setVersion11(value) {
    this._isVersion11 = value;
  }

  setVersion12(value) {
    this._isVersion12 = value;
  }

  IsNotNumericFunction() {
    const lt1 = this.getTokenStream().LT(1);
    const lt2 = this.getTokenStream().LT(2);
    if ((lt1.type === PlSqlLexer.SUM ||
      lt1.type === PlSqlLexer.COUNT ||
      lt1.type === PlSqlLexer.AVG ||
      lt1.type === PlSqlLexer.MIN ||
      lt1.type === PlSqlLexer.MAX ||
      lt1.type === PlSqlLexer.ROUND ||
      lt1.type === PlSqlLexer.LEAST ||
      lt1.type === PlSqlLexer.GREATEST) && lt2.type === PlSqlLexer.LEFT_PAREN)
      return false;
    return true;
  }

  isNotStartOfJoin() {
    const lt1 = this.getTokenStream().LT(1);
    if (lt1.type == PlSqlLexer.INNER ||
        lt1.type == PlSqlLexer.CROSS ||
        lt1.type == PlSqlLexer.NATURAL ||
        lt1.type == PlSqlLexer.PARTITION ||
        lt1.type == PlSqlLexer.FULL ||
        lt1.type == PlSqlLexer.LEFT ||
        lt1.type == PlSqlLexer.RIGHT ||
        lt1.type == PlSqlLexer.OUTER)
        return false;
    return true;
  }
}

File: sql/plsql/Python3/PlSqlParserBase.py
from antlr4 import *

class PlSqlParserBase(Parser):

    _isVersion10 = False
    _isVersion11 = True
    _isVersion12 = True

    def __init__(self, input, output=None):
        super().__init__(input, output)
        self._lastUnitWasPlsql = False

    def reset(self):
        self._lastUnitWasPlsql = False
        super().reset()

    def setLastUnitPlsql(self):
        self._lastUnitWasPlsql = True

    def setLastUnitSql(self):
        self._lastUnitWasPlsql = False

    def isLastUnitSql(self):
        return not self._lastUnitWasPlsql

    def isLastUnitPlsql(self):
        return self._lastUnitWasPlsql

    def isSolidusSeparator(self):
        from PlSqlLexer import PlSqlLexer as _Lexer
        solidus = self._input.LT(1)
        if solidus is None or solidus.type != _Lexer.SOLIDUS:
            return False

        solidusLine = solidus.line

        prev = self._input.LT(-1)
        if prev is not None and prev.type != Token.EOF and prev.line == solidusLine:
            return False

        next = self._input.LT(2)
        if next is not None and next.type != Token.EOF and next.line == solidusLine:
            return False

        return True

    def isVersion10(self):
        return self._isVersion10

    def isVersion11(self):
        return self._isVersion11

    def isVersion12(self):
        return self._isVersion12

    def setVersion10(self, value):
        self._isVersion10 = value

    def setVersion11(self, value):
        self._isVersion11 = value

    def setVersion12(self, value):
        self._isVersion12 = value

    def IsNotNumericFunction(self):
        from PlSqlLexer import PlSqlLexer as _Lexer
        lt1 = self._input.LT(1)
        lt2 = self._input.LT(2)
        if (lt1.type in (_Lexer.SUM, _Lexer.COUNT, _Lexer.AVG,
                         _Lexer.MIN, _Lexer.MAX, _Lexer.ROUND,
                         _Lexer.LEAST, _Lexer.GREATEST) and
                lt2.type == _Lexer.LEFT_PAREN):
            return False
        return True

    def isNotStartOfJoin(self):
        from PlSqlLexer import PlSqlLexer as _Lexer
        lt1 = self._input.LT(1)
        if (lt1.type in (_Lexer.INNER, _Lexer.CROSS, _Lexer.NATURAL,
                         _Lexer.PARTITION, _Lexer.FULL, _Lexer.LEFT,
                         _Lexer.RIGHT, _Lexer.OUTER)):
            return False
        return True
